### PR TITLE
nautilus: os/bluestore: apply shared_alloc_size to shared device with log level change

### PR DESCRIPTION
**Pull request by vumrao**
https://tracker.ceph.com/issues/41340
https://tracker.ceph.com/issues/41531

### CODE REVIEW

**Comment by vumrao** [comment]
cc @neha-ojha

**Comment by yuriw** [comment]
@vumrao cant merge

--- pr 30229 --- pulling https://github.com/vumrao/ceph.git branch wip-vumrao-bluefs-shared-alloc-with-log-level-change-nautilus
remote: Enumerating objects: 38, done.
remote: Counting objects: 100% (38/38), done.
remote: Total 45 (delta 38), reused 38 (delta 38), pack-reused 7
Unpacking objects: 100% (45/45), done.
From https://github.com/vumrao/ceph
 * branch                  wip-vumrao-bluefs-shared-alloc-with-log-level-change-nautilus -> FETCH_HEAD
Auto-merging src/os/bluestore/BlueStore.cc
Auto-merging src/os/bluestore/BlueFS.h
Auto-merging src/os/bluestore/BlueFS.cc
CONFLICT (content): Merge conflict in src/os/bluestore/BlueFS.cc
Auto-merging src/common/options.cc
Auto-merging src/common/legacy_config_opts.h
Automatic merge failed; fix conflicts and then commit the result.
Traceback (most recent call last):
  File "/home/yuriw/wip_master/src/script/build-integration-branch", line 67, in <module>
    assert not r
AssertionError

**Comment by vumrao** [comment]
@yuriw rebased it.

**Comment by yuriw** [comment]
[wip-yuri6-testing-2019-11-15-1956-nautilus](https://trello.com/c/geHexz7D/764-wip-yuri6-testing-2019-11-15-1956-nautilus)

**Comment by neha-ojha** [comment]
Seeing consistent failures like http://pulpito.ceph.com/yuriw-2019-11-16_03:19:41-rados-wip-yuri6-testing-2019-11-15-1956-nautilus-distro-basic-smithi/4512196/

2019-11-16T03:39:42.690 INFO:teuthology.orchestra.run.smithi062:> CEPH_REF=master CEPH_ID="0" PATH=$PATH:/usr/sbin adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage /home/ubuntu/cephtest/virtualenv/bin/cram -v -- /home/ubuntu/cephtest/archive/cram.client.0/*.t
2019-11-16T03:40:08.864 INFO:tasks.ceph.osd.1.smithi062.stderr:2019-11-16 03:40:08.865 7f9862e71700 -1 bluestore(/var/lib/ceph/osd/ceph-1) _do_alloc_write failed to allocate 0x400000 allocated 0x 384000 min_alloc_size 0x4000 available 0x 0
2019-11-16T03:40:08.864 INFO:tasks.ceph.osd.1.smithi062.stderr:2019-11-16 03:40:08.865 7f9862e71700 -1 bluestore(/var/lib/ceph/osd/ceph-1) _do_write _do_alloc_write failed with (28) No space left on device
2019-11-16T03:40:08.864 INFO:tasks.ceph.osd.1.smithi062.stderr:2019-11-16 03:40:08.865 7f9862e71700 -1 bluestore(/var/lib/ceph/osd/ceph-1) _txc_add_transaction error (28) No space left on device not handled on operation 10 (op 3, counting from 0)
2019-11-16T03:40:08.864 INFO:tasks.ceph.osd.1.smithi062.stderr:2019-11-16 03:40:08.865 7f9862e71700 -1 bluestore(/var/lib/ceph/osd/ceph-1) ENOSPC from bluestore, misconfigured cluster
2019-11-16T03:40:08.864 INFO:tasks.ceph.osd.1.smithi062.stderr:2019-11-16 03:40:08.866 7f9863672700 -1 bluestore(/var/lib/ceph/osd/ceph-1) _do_alloc_write failed to allocate 0x400000 allocated 0x 384000 min_alloc_size 0x4000 available 0x 0
2019-11-16T03:40:08.865 INFO:tasks.ceph.osd.1.smithi062.stderr:2019-11-16 03:40:08.866 7f9863672700 -1 bluestore(/var/lib/ceph/osd/ceph-1) _do_write _do_alloc_write failed with (28) No space left on device
2019-11-16T03:40:08.865 INFO:tasks.ceph.osd.1.smithi062.stderr:2019-11-16 03:40:08.866 7f9863672700 -1 bluestore(/var/lib/ceph/osd/ceph-1) _txc_add_transaction error (28) No space left on device not handled on operation 10 (op 3, counting from 0)
2019-11-16T03:40:08.865 INFO:tasks.ceph.osd.1.smithi062.stderr:2019-11-16 03:40:08.866 7f9863672700 -1 bluestore(/var/lib/ceph/osd/ceph-1) ENOSPC from bluestore, misconfigured cluster
2019-11-16T03:40:08.867 INFO:tasks.ceph.osd.1.smithi062.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.2.4-660-gd1576af/rpm/el7/BUILD/ceph-14.2.4-660-gd1576af/src/os/bluestore/BlueStore.cc: In function 'void BlueStore::_txc_add_transaction(BlueStore::TransContext*, ObjectStore::Transaction*)' thread 7f9862e71700 time 2019-11-16 03:40:08.866489
2019-11-16T03:40:08.867 INFO:tasks.ceph.osd.1.smithi062.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.2.4-660-gd1576af/rpm/el7/BUILD/ceph-14.2.4-660-gd1576af/src/os/bluestore/BlueStore.cc: 11319: ceph_abort_msg("unexpected error")
2019-11-16T03:40:08.867 INFO:tasks.ceph.osd.1.smithi062.stderr: ceph version 14.2.4-660-gd1576af (d1576af8d2be2b66b208bdc8b26056352779119d) nautilus (stable)
2019-11-16T03:40:08.867 INFO:tasks.ceph.osd.1.smithi062.stderr: 1: (ceph::__ceph_abort(char const*, int, char const*, std::string const&)+0xdd) [0x562e21cb8789]
2019-11-16T03:40:08.867 INFO:tasks.ceph.osd.1.smithi062.stderr: 2: (BlueStore::_txc_add_transaction(BlueStore::TransContext*, ObjectStore::Transaction*)+0xc63) [0x562e221da7c3]
2019-11-16T03:40:08.867 INFO:tasks.ceph.osd.1.smithi062.stderr: 3: (BlueStore::queue_transactions(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, std::vector<ObjectStore::Transaction, std::allocator<ObjectStore::Transaction> >&, boost::intrusive_ptr<TrackedOp>, ThreadPool::TPHandle*)+0x528) [0x562e221e2708]
2019-11-16T03:40:08.868 INFO:tasks.ceph.osd.1.smithi062.stderr: 4: (non-virtual thunk to PrimaryLogPG::queue_transactions(std::vector<ObjectStore::Transaction, std::allocator<ObjectStore::Transaction> >&, boost::intrusive_ptr<OpRequest>)+0x54) [0x562e21f43564]
2019-11-16T03:40:08.868 INFO:tasks.ceph.osd.1.smithi062.stderr: 5: (ReplicatedBackend::do_repop(boost::intrusive_ptr<OpRequest>)+0xaa9) [0x562e2203f0c9]
2019-11-16T03:40:08.868 INFO:tasks.ceph.osd.1.smithi062.stderr: 6: (ReplicatedBackend::_handle_message(boost::intrusive_ptr<OpRequest>)+0x187) [0x562e2204d527]
2019-11-16T03:40:08.868 INFO:tasks.ceph.osd.1.smithi062.stderr: 7: (PGBackend::handle_message(boost::intrusive_ptr<OpRequest>)+0x87) [0x562e21f551d7]
2019-11-16T03:40:08.868 INFO:tasks.ceph.osd.1.smithi062.stderr: 8: (PrimaryLogPG::do_request(boost::intrusive_ptr<OpRequest>&, ThreadPool::TPHandle&)+0x695) [0x562e21f00cd5]
2019-11-16T03:40:08.868 INFO:tasks.ceph.osd.1.smithi062.stderr: 9: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x1a9) [0x562e21d38e99]
2019-11-16T03:40:08.868 INFO:tasks.ceph.osd.1.smithi062.stderr: 10: (PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x62) [0x562e21fdac72]
2019-11-16T03:40:08.869 INFO:tasks.ceph.osd.1.smithi062.stderr: 11: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x9f4) [0x562e21d56d24]
2019-11-16T03:40:08.869 INFO:tasks.ceph.osd.1.smithi062.stderr: 12: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x433) [0x562e2232b123]
2019-11-16T03:40:08.869 INFO:tasks.ceph.osd.1.smithi062.stderr: 13: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x562e2232e1c0]
2019-11-16T03:40:08.869 INFO:tasks.ceph.osd.1.smithi062.stderr: 14: (()+0x7ea5) [0x7f988965fea5]
2019-11-16T03:40:08.869 INFO:tasks.ceph.osd.1.smithi062.stderr: 15: (clone()+0x6d) [0x7f98885238cd]
2019-11-16T03:40:08.869 INFO:tasks.ceph.osd.1.smithi062.stderr:*** Caught signal (Aborted) **

**Comment by vumrao** [comment]
@neha-ojha I have rebased but looks like code changed in BlueFS.cc can you please review. Thank you.

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -510,21 +509,38 @@ void BlueFS::_init_alloc()
...
+  if (bdev[BDEV_NEWDB]) {
+    alloc_size[BDEV_NEWDB] = cct->_conf->bluefs_alloc_size;
+  }
+
   for (unsigned id = 0; id < bdev.size(); ++id) {
     if (!bdev[id]) {
       continue;
     }
     ceph_assert(bdev[id]->get_size());
-    std::string name = "bluefs-";
```
Comment (vumrao) [inline src/os/bluestore/BlueFS.cc]: @neha-ojha This was the code which was causing conflict ... looks like recently merged in nautilus.

File: src/os/bluestore/BlueFS.cc [hunk 2]
```diff
@@ -510,21 +509,38 @@ void BlueFS::_init_alloc()
...
-    const char* devnames[] = {"wal","db","slow"};
-    if (id <= BDEV_SLOW)
-      name += devnames[id];
-    else
-      name += to_string(uintptr_t(this));
```
Comment (vumrao) [inline src/os/bluestore/BlueFS.cc]: @neha-ojha This was the code that came recent and was causing conflict.

**Comment by vumrao** [comment]
> @neha-ojha I have rebased but looks like code changed in BlueFS.cc can you please review. Thank you.

I think I need to clean this up first. As a lot has changed in PendingRelease notes etc ... I need to look carefully will do it and let you know ...

**Comment by vumrao** [comment]
> > @neha-ojha I have rebased but looks like code changed in BlueFS.cc can you please review. Thank you.
> 
> I think I need to clean this up first. As a lot has changed in PendingRelease notes etc ... I need to look carefully will do it and let you know ...

@neha-ojha - I have done a fresh rebase and backport on the latest nautilus and I think it should be fine now. Thank you. Can you please take a look ...

**Comment by yuriw** [comment]
[wip-yuri4-testing-2019-11-19-2209-nautilus](https://trello.com/c/7N3mUNdY/770-wip-yuri4-testing-2019-11-19-2209-nautilus)

**Comment by vumrao** [comment]
Again same:

~~~
019-11-20T01:37:39.571 INFO:teuthology.orchestra.run.smithi040.stderr:HEAD is now at d418147728 Merge branch 'wip-vumrao-bluefs-shared-alloc-with-log-level-change-nautilus' of https://github.com/vumrao/ceph into wip-yuri4-testing-2019-11-19-2209
2019-11-20T01:37:39.572 INFO:teuthology.orchestra.run.smithi040:> cp -- /home/ubuntu/cephtest/clone.client.0/src/test/cli-integration/balancer/misplaced.t /home/ubuntu/cephtest/archive/cram.client.0
2019-11-20T01:37:39.581 INFO:tasks.cram:Running tests for client.0...
2019-11-20T01:37:39.581 INFO:teuthology.orchestra.run.smithi040:> CEPH_REF=master CEPH_ID="0" PATH=$PATH:/usr/sbin adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage /home/ubuntu/cephtest/virtualenv/bin/cram -v -- /home/ubuntu/cephtest/archive/cram.client.0/*.t
2019-11-20T01:38:04.220 INFO:tasks.ceph.osd.1.smithi040.stderr:2019-11-20 01:38:04.215 7fb3948e9700 -1 bluestore(/var/lib/ceph/osd/ceph-1) _do_alloc_write failed to allocate 0x400000 allocated 0x 384000 min_alloc_size 0x4000 available 0x 0
2019-11-20T01:38:04.220 INFO:tasks.ceph.osd.1.smithi040.stderr:2019-11-20 01:38:04.215 7fb3948e9700 -1 bluestore(/var/lib/ceph/osd/ceph-1) _do_write _do_alloc_write failed with (28) No space left on device
2019-11-20T01:38:04.221 INFO:tasks.ceph.osd.1.smithi040.stderr:2019-11-20 01:38:04.215 7fb3948e9700 -1 bluestore(/var/lib/ceph/osd/ceph-1) _txc_add_transaction error (28) No space left on device not handled on operation 10 (op 3, counting from 0)
2019-11-20T01:38:04.221 INFO:tasks.ceph.osd.1.smithi040.stderr:2019-11-20 01:38:04.215 7fb3948e9700 -1 bluestore(/var/lib/ceph/osd/ceph-1) ENOSPC from bluestore, misconfigured cluster
2019-11-20T01:38:04.224 INFO:tasks.ceph.osd.1.smithi040.stderr:/build/ceph-14.2.4-858-gd418147/src/os/bluestore/BlueStore.cc: In function 'void BlueStore::_txc_add_transaction(BlueStore::TransContext*, ObjectStore::Transaction*)' thread 7fb3948e9700 time 2019-11-20 01:38:04.220903
2019-11-20T01:38:04.224 INFO:tasks.ceph.osd.1.smithi040.stderr:/build/ceph-14.2.4-858-gd418147/src/os/bluestore/BlueStore.cc: 11378: ceph_abort_msg("unexpected error")
2019-11-20T01:38:04.224 INFO:tasks.ceph.osd.1.smithi040.stderr: ceph version 14.2.4-858-gd418147 (d4181477281bbed7fdbf9c223118c1b88e72008b) nautilus (stable)
2019-11-20T01:38:04.225 INFO:tasks.ceph.osd.1.smithi040.stderr: 1: (ceph::__ceph_abort(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xdf) [0x55fbc1ba4f76]
2019-11-20T01:38:04.225 INFO:tasks.ceph.osd.1.smithi040.stderr: 2: (BlueStore::_txc_add_transaction(BlueStore::TransContext*, ObjectStore::Transaction*)+0xced) [0x55fbc20efccd]
2019-11-20T01:38:04.225 INFO:tasks.ceph.osd.1.smithi040.stderr: 3: (BlueStore::queue_transactions(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, std::vector<ObjectStore::Transaction, std::allocator<ObjectStore::Transaction> >&, boost::intrusive_ptr<TrackedOp>, ThreadPool::TPHandle*)+0x2aa) [0x55fbc20f5bba]
2019-11-20T01:38:04.225 INFO:tasks.ceph.osd.1.smithi040.stderr: 4: (non-virtual thunk to PrimaryLogPG::queue_transactions(std::vector<ObjectStore::Transaction, std::allocator<ObjectStore::Transaction> >&, boost::intrusive_ptr<OpRequest>)+0x54) [0x55fbc1e43c84]
2019-11-20T01:38:04.226 INFO:tasks.ceph.osd.1.smithi040.stderr: 5: (ReplicatedBackend::do_repop(boost::intrusive_ptr<OpRequest>)+0xb09) [0x55fbc1f45019]
2019-11-20T01:38:04.226 INFO:tasks.ceph.osd.1.smithi040.stderr: 6: (ReplicatedBackend::_handle_message(boost::intrusive_ptr<OpRequest>)+0x1a7) [0x55fbc1f533a7]
2019-11-20T01:38:04.226 INFO:tasks.ceph.osd.1.smithi040.stderr: 7: (PGBackend::handle_message(boost::intrusive_ptr<OpRequest>)+0x97) [0x55fbc1e56fe7]
2019-11-20T01:38:04.226 INFO:tasks.ceph.osd.1.smithi040.stderr: 8: (PrimaryLogPG::do_request(boost::intrusive_ptr<OpRequest>&, ThreadPool::TPHandle&)+0x705) [0x55fbc1e05775]
2019-11-20T01:38:04.227 INFO:tasks.ceph.osd.1.smithi040.stderr: 9: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x1bf) [0x55fbc1c2794f]
2019-11-20T01:38:04.227 INFO:tasks.ceph.osd.1.smithi040.stderr: 10: (PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x62) [0x55fbc1ede702]
2019-11-20T01:38:04.227 INFO:tasks.ceph.osd.1.smithi040.stderr: 11: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0xbf5) [0x55fbc1c46425]
2019-11-20T01:38:04.227 INFO:tasks.ceph.osd.1.smithi040.stderr: 12: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x4ac) [0x55fbc2245d1c]
2019-11-20T01:38:04.228 INFO:tasks.ceph.osd.1.smithi040.stderr: 13: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x55fbc2248ed0]
2019-11-20T01:38:04.228 INFO:tasks.ceph.osd.1.smithi040.stderr: 14: (()+0x76db) [0x7fb3ba43d6db]
2019-11-20T01:38:04.228 INFO:tasks.ceph.osd.1.smithi040.stderr: 15: (clone()+0x3f) [0x7fb3b91dd88f]
~~~

- I see this code was merged recently in BlueFS - https://github.com/ceph/ceph/pull/30229/files#diff-6d10a1d221da60f4fb9985631ce2125cR537-R542, maybe related for this issue?

**Comment by neha-ojha** [comment]
@vumrao That failure does not seem related to this PR. It has appeared in other batches since we merged https://github.com/ceph/ceph/pull/31682/commits/15f360dd17502b947c0237fd6b01a8b14e9be6c7. For example see the dead job in http://pulpito.ceph.com/yuriw-2019-11-19_20:27:29-rados-wip-yuri3-testing-2019-11-19-1618-nautilus-distro-basic-smithi/
I'll create a tracker issue to investigate that separately.

**Pull request merged**